trax/tf_numpy/examples/mnist/model.py

```diff
@@ -19,119 +19,109 @@
-
-  def forward(self, x):
-    """Performs the forward pass.
-
-    Args:
-      x: 2-d array of size batch_size x image_size.
-
-    Returns:
-      A 2-d array of size batch_size x num_classes.
     """
 
-    def sigmoid(x):
-      return 1.0 / (1.0 + np.exp(-x))
+    def __init__(self, hidden_layers, input_size=784, num_classes=10, learning_rate=0.001):
+        """Initializes the neural network.
 
-    for w, b in zip(self.weights, self.biases):
-      x = sigmoid(np.dot(w, x.T).T + b)
-    return x
+        Args:
+          hidden_layers: List of ints specifying the sizes of hidden layers. Could
+            be empty.
+          input_size: Length of the input array. The network receives the input
+            image as a flattened 1-d array. Defaults to 784(28*28), the default
+            image size for MNIST.
+          num_classes: The number of output classes. Defaults to 10.
+        """
+        hidden_layers = [input_size] + hidden_layers + [num_classes]
+        self.params = []
+        for i in range(len(hidden_layers) - 1):
+            self.params.append(np.random.randn(hidden_layers[i + 1], hidden_layers[i]))
+            self.params.append(np.random.randn(hidden_layers[i + 1]))
 
-  def train(self, x, y, learning_rate=0.01):
-    """Runs a single training pass.
 
-    Args:
-      x: 2-d array of size batch_size x image_size.
-      y: 2-d array of size batch_size x num_classes in one-hot notation.
-      learning_rate: The learning rate.
-    """
-    x = np.array(x, copy=False)
-    y = np.array(y, copy=False)
+    def sigmoid(self, x):
+        return 1 / (1 + np.exp(-x))
 
-    def mean_squared_error(x, y):
-      diff = x - y
-      return np.sum(diff * diff) / len(x)
 
-    wb_tensors = [p.data for p in self.weights + self.biases]
-    with tf.GradientTape() as g:
-      g.watch(wb_tensors)
-      loss = mean_squared_error(self.forward(x), y)
-    gradients = g.gradient(loss.data, wb_tensors)
-    gradients = [np.asarray(grad) for grad in gradients]
+    def forward(self, x):
+        """Performs the forward pass.
 
-    new_weights_and_biases = []
-    for v, dv in zip(self.weights + self.biases, gradients):
-      new_weights_and_biases.append(v - learning_rate * dv)
+        Args:
+          x: 2-d array of size batch_size x image_size.
 
-    total_len = len(new_weights_and_biases)
-    self.weights = new_weights_and_biases[:total_len // 2]
-    self.biases = new_weights_and_biases[total_len // 2:]
+        Returns:
+          A 2-d array of size batch_size x num_classes.
+        """
 
-  def evaluate(self, x, y):
-    """Returns the number of correct predictions.
+        out = tf.cast(x, tf.float32)
 
-    Args:
-      x: 2-d array of size batch_size x image_size.
-      y: 2-d array of size batch_size x num_classes.
+        for i in range(0, len(self.params)-2, 2):
+            out = np.tanh(np.dot(out, self.params[i].T) + self.params[i+1])
```

Comment:
A few questions about this:
- The params seem to be transposed every time - it seems like we're doing initializing with randn(layer+1.size, layer.size). Can't we just switch that and you won't need the transpose?
- This was changed from sigmoid to tanh for the hidden layers - was there some particular reason for that?

Reply:
1. Yes that makes more sense, I have now switched them.
2. Tanh tends to be a better midlayer to use because it is centered around 0 and also has larger derivatives. Also np does not include a native sigmoid so np.tanh is probably numerically stable than a self built sigmoid.